src/sound_distance.py

```diff
@@ -0,0 +1,115 @@
+
+        self.all_sounds = set(list(self.all_vowels) + list(self.all_consonants))
+        self.all_properties = set(list(self.vowel_properties) + list(self.consonant_properties))
+
+        self.data = {}
+        for consonant in self.all_consonants:
+            self.data[consonant]={}
+            weights = 0
+            for property in self.consonant_properties:
+                if consonant in self.consonant_data[property].split():
+                    weights += self.consonant_data['properties'][property]**2
+                    self.data[consonant][property] = self.consonant_data['properties'][property]
+                else:
+                    self.data[consonant][property] = 0
+            #  special treatment for voiced/voiceless axis
+            voicedness = self.data[consonant]['voiced']-self.data[consonant]['voiceless']
+            self.data[consonant]['voiced'], self.data[consonant]['voiceless'] = voicedness, -voicedness
+            weights += self.consonant_data['properties']['voiced'] ** 2
+
+            for property in self.consonant_properties:  # normalization
+                self.data[consonant][property] /= sqrt(weights)
+
+            for property in self.vowel_properties:
+                self.data[consonant][property] = 0
+
+        for vowel in self.all_vowels:
+            self.data[vowel] = {}
+            weights = 0
+            for property in self.vowel_properties:
+                self.data[vowel][property] = self.vowel_data[property][vowel]*self.vowel_data['properties'][property]
+                weights += self.data[vowel][property]**2
+
+            for property in self.vowel_properties:  # normalization
+                self.data[vowel][property] /= sqrt(weights)
+
+            for property in self.consonant_properties:
+                self.data[vowel][property] = 0
+
+    def detect_sounds(self, str):
+        """
+        Parses a string to detect known sound symbols
+        Necessary because of 0-space chars in sounds (diactitics etc.)
+        :param str:
+        :return:
+        """
+        i = len(str)
+        if i == 0:
+            return []
+        while i > 0:
+            if str[:i] in self.all_sounds:
+                return [str[:i]] + self.detect_sounds(str[i:])
+            else:
+                i -= 1
+
+    def sound_similarity(self, s1, s2):
+        res = 0
+        for property in self.all_properties:
+            res += self.data[s1][property]*self.data[s2][property]
+        return res
+
+    def syllable_similarity(self, s1, s2):
+        res = 0
+        weights = 0
+        sounds1, sounds2 = self.detect_sounds(s1), self.detect_sounds(s2)
+        if len(sounds1) == len(sounds2):
+            for i in range(len(sounds1)):
+                s = self.sound_similarity(sounds1[i], sounds2[i])
+                res += s
+                weights += 1
+            if weights > 0:
+                res /= weights
+        return res
+
+
+if __name__ == '__main__':
```

Comment:
better to put that in another file IMO

Reply:
These are my own little test, we still have to discuss where exactly this will be used (as a batch calculation on every single syllable or on the fly for each request, more likely the first one)

Reply:
if those are tests... then make it a unit test !! look at unittest it part of the python  distribution. it's very easy and very standard.